src/main/java/picard/sam/CreateSequenceDictionary.java

```diff
@@ -134,33 +132,53 @@ protected int doWork() {
             throw new PicardException(OUTPUT.getAbsolutePath() +
                     " already exists.  Delete this file and try again, or specify a different output file.");
         }
-        final SAMSequenceDictionary sequences = makeSequenceDictionary(REFERENCE);
-        final SAMFileHeader samHeader = new SAMFileHeader();
-        samHeader.setSequenceDictionary(sequences);
-        final SAMFileWriter samWriter = new SAMFileWriterFactory().makeSAMWriter(samHeader, false, OUTPUT);
-        samWriter.close();
-        return 0;
-    }
 
-    /**
-     * Read all the sequences from the given reference file, and convert into SAMSequenceRecords
-     * @param referenceFile fasta or fasta.gz
-     * @return SAMSequenceRecords containing info from the fasta, plus from cmd-line arguments.
-     */
-    public SAMSequenceDictionary makeSequenceDictionary(final File referenceFile) {
-        final ReferenceSequenceFile refSeqFile =
-                ReferenceSequenceFileFactory.getReferenceSequenceFile(referenceFile, TRUNCATE_NAMES_AT_WHITESPACE);
-        ReferenceSequence refSeq;
-        final List<SAMSequenceRecord> ret = new ArrayList<SAMSequenceRecord>();
-        final Set<String> sequenceNames = new HashSet<String>();
-        for (int numSequences = 0; numSequences < NUM_SEQUENCES && (refSeq = refSeqFile.nextSequence()) != null; ++numSequences) {
-            if (sequenceNames.contains(refSeq.getName())) {
-                throw new PicardException("Sequence name appears more than once in reference: " + refSeq.getName());
+        // SortingCollection is used to check uniqueness of sequence names
+        final SortingCollection<String> sequenceNames = makeSortingCollection();
+        try (BufferedWriter writer = makeWriter()) {
+            final ReferenceSequenceFile refSeqFile = ReferenceSequenceFileFactory.
+                    getReferenceSequenceFile(REFERENCE, TRUNCATE_NAMES_AT_WHITESPACE);
+            SAMSequenceDictionaryCodec samDictCodec = new SAMSequenceDictionaryCodec(writer);
+
+            samDictCodec.encodeHeaderLine(false);
+            // read reference sequence one by one and write its metadata
+            ReferenceSequence refSeq;
+            while ((refSeq = refSeqFile.nextSequence()) != null) {
```

Comment:
ReferenceSequence is not Iterable. It doesn't have any methods to get some Iterable (f.e. FastaSequenceIndex). One possible solution is to make it Iterable. Another way is to write for loop in the following way (which is worse): 
````
for (ReferenceSequence refSeq = refSeqFile.nextSequence(); refSeq != null; refSeq = refSeqFile.nextSequence()) {
       final SAMSequenceRecord samSequenceRecord = makeSequenceRecord(refSeq);
       samDictCodec.encodeSequenceRecord(samSequenceRecord);
       sequenceNames.add(refSeq.getName());
}
````